keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,730 @@
++    // hugepages).
++    UInt64Value rss_bytes = 5;
++    // Cumulative number of minor page faults.
++    Uint64Value page_faults = 6;
++    // Cumulative number of major page faults.
++    Uint64Value major_page_faults = 6;
+-}
+
+-// FilesystemUsage provides the filesystem usage information.
+-message FilesystemUsage {
+-    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
++    // Corresponds to Stats Summary API FsStats Time field
+-    int64 timestamp = 1;
+-    // The unique identifier of the filesystem.
++    // Does not correspond to any field in Stats Summary API FsStats
+-    FilesystemIdentifier fs_id = 2;
+-    // UsedBytes represents the bytes used for images on the filesystem.
+-    // This may differ from the total bytes used on the filesystem and may not
+-    // equal CapacityBytes - AvailableBytes.
+-    // Corresponds to Stats Summary API FsStats UsedBytes field
+-    UInt64Value used_bytes = 3;
+-    // InodesUsed represents the inodes used by the images.
+-    // This may not equal InodesCapacity - InodesAvailable because the underlying
+-    // filesystem may also be used for purposes other than storing images.
+-    // Corresponds to Stats Summary API FsStats InodesUsed field
+-    UInt64Value inodes_used = 4;
++    // TODO: Unclear how the remaining fields relate to container stats. Is it filled in cAdvisor?
++    // AvailableBytes represents the storage space available (bytes) for the filesystem.
++    UInt64Value available_bytes = 5;
++    // CapacityBytes represents the total capacity (bytes) of the filesystems underlying storage.
++    UInt64Value capacity_bytes = 6;
++    // InodesFree represents the free inodes in the filesystem.
++    UInt64Value inodes_free = 7;
++    // Inodes represents the total inodes in the filesystem.
++    UInt64Value inodes = 8;
+-}
+```
+
+All together, below is the proposition for the new `ContainerStats` object:
+```
+-// ContainerStats provides the resource usage statistics for a container.
+-message ContainerStats {
+-    // Information of the container.
++    // Corresponds to the Stats Summary API ContainerStats Name field
+-    ContainerAttributes attributes = 1;
+-    // CPU usage gathered from the container.
++    // Corresponds to Stats Summary API ContainerStats CPUStats field
+-    CpuUsage cpu = 2;
+-    // Memory usage gathered from the container.
++    // Corresponds to Stats Summary API ContainerStats MemoryStats field
+-    MemoryUsage memory = 3;
+-    // Usage of the writable layer.
++    // Corresponds to Stats Summary API ContainerStats Rootfs field
+-    FilesystemUsage writable_layer = 4;
++    // Stats pertaining to container logs usage of filesystem resources
++    // Logs.UsedBytes is the number of bytes used for the container logs.
++    FilesystemUsage logs = 5;
+-}
+```
+**TODO**: in Stats Summary API ContainerStats object, there's a time field. Do we want to add this to `ContainerStats`?
```

Comment:
Who needs to weigh in on this?

Reply:
@dashpole already did but I forgot to add it. I think we do want this, as it allows CRI impls to have more flexibility about when they scrape the metrics (gathering on their own time rather than on demand)

Reply:
One potential implication is for autoscaling.  Requiring on-demand collection would ensure the autoscaling experience is consistent across clusters.  I know of some use-cases where they do 1s granularity scraping of the summary API or resource metrics API to do faster autoscaling.  If the metrics are being collected at an interval (probable 10s or so), that would break such pipelines.

We could decide to treat different resources differently... E.g. only add timestamps for disk metrics because collecting them is quite expensive.

Reply:
I like the idea of only doing it for FS. we'd need buy in from other CRI impls 
cc @mikebrow @dims @crosbymichael

Reply:
+1 from me ... also cc @saschagrunert @mrunalp

Reply:
> I like the idea of only doing it for FS. we'd need buy in from other CRI impls

I think we do on-demand cpu/memory stats collection, but periodical disk stats collection in containerd.

Reply:
Please update this based on the above conversation. I think only having the timestamp only for FS is acceptable. btw disk related stats is periodically collected by design, but collecting them might take longer time in some circumstances, which makes the timestamp more important here.

Reply:
I removed the TODO and added a blurb about how each individual struct has a timestamp, allowing for more flexibility of when cri impl fetches what